File: Java/src/SubsetsIITest.java
import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;

public class SubsetsIITest {

    @Test
    public void testSubsetsWithDup() {
        SubsetsII s = new SubsetsII();

        Assert.assertTrue(s.subsetsWithDup(null).isEmpty());

        Assert.assertTrue(s.subsetsWithDup(new int[]{}).isEmpty());

        Assert.assertEquals(s.subsetsWithDup(new int[]{1}),
                Arrays.asList(Collections.emptyList(), Collections.singletonList(1)));

        Assert.assertEquals(new HashSet<>(s.subsetsWithDup(new int[]{1, 1})),
                new HashSet<>(Arrays.asList(Collections.emptyList(), Collections.singletonList(1), Arrays.asList(1, 1))));

        Assert.assertEquals(new HashSet<>(s.subsetsWithDup(new int[]{1, 2})),
                new HashSet<>(Arrays.asList(Collections.emptyList(), Collections.singletonList(1), Collections.singletonList(2), Arrays.asList(1, 2))));

        Assert.assertEquals(new HashSet<>(s.subsetsWithDup(new int[]{1, 2, 2})),
                new HashSet<>(Arrays.asList(Collections.emptyList(), Collections.singletonList(1), Collections.singletonList(2), Arrays.asList(1, 2), Arrays.asList(2, 2), Arrays.asList(1, 2, 2))));

        Assert.assertEquals(new HashSet<>(s.subsetsWithDup(new int[]{4, 4, 4, 1, 4})),
                new HashSet<>(Arrays.asList(Collections.emptyList(), Collections.singletonList(1), Collections.singletonList(4), Arrays.asList(1, 4), Arrays.asList(1, 4, 4), Arrays.asList(1, 4, 4, 4), Arrays.asList(1, 4, 4, 4, 4), Arrays.asList(4, 4), Arrays.asList(4, 4, 4), Arrays.asList(4, 4, 4, 4))));
    }

}

File: Kotlin/src/SubsetsIITest.kt
import org.junit.Test

class SubsetsIITest {

    @Test
    fun testSubsetsWithDup() {
        val s = SubsetsII()

        assert(s.subsetsWithDup(intArrayOf()).isEmpty())

        val list0 = s.subsetsWithDup(intArrayOf(1))
        assert(list0.size == 2)
        assert(list0.toSet() == listOf(listOf(), listOf(1)).toSet())

        val list1 = s.subsetsWithDup(intArrayOf(1, 1))
        assert(list1.size == 3)
        assert(list1.toSet() == listOf(listOf(), listOf(1), listOf(1, 1)).toSet())

        val list2 = s.subsetsWithDup(intArrayOf(1, 2))
        assert(list2.size == 4)
        assert(list2.toSet() == listOf(listOf(), listOf(1), listOf(2), listOf(2, 1)).toSet())

        val list3 = s.subsetsWithDup(intArrayOf(1, 2, 2))
        assert(list3.size == 6)
        assert(list3.toSet() == listOf(listOf(), listOf(1), listOf(2), listOf(2, 2, 1), listOf(2, 2), listOf(2, 1)).toSet())

        val list4 = s.subsetsWithDup(intArrayOf(4, 4, 4, 1, 4))
        assert(list4.size == 10)
        assert(list4.toSet() == listOf(listOf(), listOf(1), listOf(4, 1), listOf(4, 4, 1), listOf(4, 4, 4, 1), listOf(4, 4, 4, 4, 1), listOf(4), listOf(4, 4), listOf(4, 4, 4), listOf(4, 4, 4, 4)).toSet())
    }

}